spec/core/ics-002-client-semantics/README.md

```diff
@@ -278,148 +227,113 @@ Client types MUST define a method to initialise a client state with a provided c
+```typescript
+type getTimestampAtHeight = (
+  clientState: ClientState,
+  height: Height
+) => uint64
+```
 
-#### State verification
+#### ClientMessage
 
-Client types must define functions to authenticate internal state of the state machine which the client tracks.
-Internal implementation details may differ (for example, a loopback client could simply read directly from the state and require no proofs).
+A `ClientMessage` is an opaque data structure defined by a client type which provides information to update the client.
+`ClientMessages` can be submitted to an associated client to add new `ConsensusState(s)` and/or update the `ClientState`. They likely contain a height, a proof, a commitment root, and possibly updates to the validity predicate.
 
-- The `delayPeriodTime` is passed to packet-related verification functions in order to allow packets to specify a period of time which must pass after a header is verified before the packet is allowed to be processed.
-- The `delayPeriodBlocks` is passed to packet-related verification functions in order to allow packets to specify a period of blocks which must pass after a header is verified before the packet is allowed to be processed.
+```typescript
+type ClientMessage = bytes
+```
 
-##### Required functions
+#### Validity predicate
+
+A validity predicate is an opaque function defined by a client type to verify `ClientMessage`s depending on the current `ConsensusState`.
+Using the validity predicate SHOULD be far more computationally efficient than replaying the full consensus algorithm
+for the given parent `ClientMessage` and the list of network messages.
 
-`verifyClientConsensusState` verifies a proof of the consensus state of the specified client stored on the target state machine.
+The validity predicate is defined as:
 
 ```typescript
-type verifyClientConsensusState = (
-  clientState: ClientState,
-  height: Height,
-  proof: CommitmentProof,
-  clientIdentifier: Identifier,
-  consensusStateHeight: Height,
-  consensusState: ConsensusState)
-  => boolean
+type VerifyClientMessage = (ClientMessage) => Void
 ```
 
-`verifyClientState` verifies a proof of the client state of the specified client stored on the target machine.
+`VerifyClientMessage` MUST throw an exception if the provided ClientMessage was not valid.
 
-```typescript
-type verifyClientState = (
-  height: Height,
-  prefix: CommitmentPrefix,
-  proof: CommitmentProof,
-  clientIdentifier: Identifier,
-  clientState: ClientState)
-  => boolean
-```
+#### Misbehaviour predicate
 
-`verifyConnectionState` verifies a proof of the connection state of the specified connection end stored on the target state machine.
+A misbehaviour predicate is an opaque function defined by a client type, used to check if a ClientMessage
+constitutes a violation of the consensus protocol. For example, if the state machine is a blockchain; this might be two signed headers
```

Comment:
nit
```suggestion
constitutes a violation of the consensus protocol. For example, if the state machine is a blockchain, this might be two signed headers
```